witchcraft/internal/tcpjson/conn_provider.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"crypto/tls"
+	"net"
+	"net/url"
+	"sync/atomic"
+
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+// ConnProvider defines the behavior to retrieve an established net.Conn.
+type ConnProvider interface {
+	// GetConn returns a net.Conn or an error if there is no connection established.
+	// It is the users responsibility to close the returned net.Conn.
+	GetConn() (net.Conn, error)
+}
+
+const (
+	ErrNoURIs           = "no URIs provided to connect to"
+	ErrFailedParsingURI = "failed to parse provided URI"
+	ErrFailedDial       = "failed to dial the host:port"
+)
+
+var _ ConnProvider = (*TCPConnProvider)(nil)
+
+// TCPConnProvider implements a ConnProvider that will round-robin TCP connections to a specific set of hosts.
+type TCPConnProvider struct {
+	hostIdx   int32
+	hosts     []string
+	tlsConfig *tls.Config
+}
+
+func NewTCPConnProvider(uris []string, tlsCfg *tls.Config) (*TCPConnProvider, error) {
+	if len(uris) < 1 {
+		return nil, werror.Error(ErrNoURIs)
+	}
+
+	// Extract the host:port from each provided URI to simplify establishing a Dial connection
+	var hosts []string
+	for _, uri := range uris {
+		u, err := url.Parse(uri)
+		if err != nil {
+			return nil, werror.Error(ErrFailedParsingURI)
+		}
+		hosts = append(hosts, u.Host)
+	}
+
+	return &TCPConnProvider{
+		hostIdx:   0,
+		hosts:     hosts,
+		tlsConfig: tlsCfg,
+	}, nil
+}
+
+func (s *TCPConnProvider) GetConn() (net.Conn, error) {
+	oldIdx := atomic.LoadInt32(&s.hostIdx)
```

Comment:
Current logic works, but is a bit confusing to follow -- specifically, it's a bit strange to me that the idx being stored is the "next" one, but before the mod operation is applied, meaning that the value in the struct as stored may actually be out of bounds until the mod operation is applied.

Think it would be cleaner to have the `s.hostIdx` be the index of the next host that will be handed out, and then apply the mod operation before updating.

If the logic is written this way to support the case where the length of `s.hosts` might change between calls then maybe the current approach makes more sense, but doubt that is the case given that there isn't any code to guarantee the safety around that slice between accesses (and if this is the case, then I think it would be helpful to have a comment to explain so).

Reply:
I see how that can be confusing and think your suggestion makes a lot of sense. Given we're currently expecting `s.hosts` to be static, I am going to update with your suggestion and always store the next index to use and perform the mod before storing. This way the index we use is just whatever the current value is.